src/Heap.sol

```diff
@@ -32,209 +36,171 @@ library BasicHeap {
     /// @param _heap The heap to modify.
-    /// @param _rank The index of the account to move.
-    function shiftDown(Heap storage _heap, uint256 _rank) private {
-        uint256 size = getSize(_heap);
-        Account memory initialAccount = getAccount(_heap, _rank);
-        uint256 initialValue = initialAccount.value;
-        Account memory childAccount;
-        uint256 childRank = _rank * 2;
-        // At this point, childRank (resp. childRank+1) is the rank of the left (resp. right) child.
-
-        while (childRank <= size) {
-            // Compute the rank of the child with largest value.
-            if (
-                childRank < size &&
-                getAccount(_heap, childRank + 1).value > getAccount(_heap, childRank).value
-            ) childRank++;
-
-            childAccount = getAccount(_heap, childRank);
-
-            if (childAccount.value > initialValue) {
-                setAccount(_heap, _rank, childAccount);
-                _rank = childRank;
-                childRank *= 2;
-            } else break;
+    /// @param _size The size of the heap.
+    /// @param _accountToShift The account to move.
+    /// @param _index The index of the account to move.
+    function shiftDown(Heap storage _heap, uint256 _size, Account memory _accountToShift, uint256 _index) private {
+        uint256 childIndex = (_index << 1) + 1;
+        uint256 rightChildIndex;
+        // At this point, childIndex (resp. childIndex+1) is the index of the left (resp. right) child.
+
+        while (childIndex < _size) {
+            Account memory childToSwap = _heap.accounts[childIndex];
+
+            // Find the child with largest value.
+            unchecked {
+                rightChildIndex = childIndex + 1; // This cannot overflow because childIndex < size.
+            }
+            if (rightChildIndex < _size) {
+                Account memory rightChild = _heap.accounts[rightChildIndex];
+                if (rightChild.value > childToSwap.value) {
+                    childToSwap = rightChild;
+                    childIndex = rightChildIndex;
+                }
+            }
+
+            if (childToSwap.value > _accountToShift.value) {
+                setAccount(_heap, childToSwap, _index);
+                _index = childIndex;
+                childIndex = (childIndex << 1) + 1;
+            } else {
+                break;
+            }
         }
-        setAccount(_heap, _rank, initialAccount);
+
+        setAccount(_heap, _accountToShift, _index);
     }
 
     /// VIEW ///
 
-    /// @notice Returns the size of the `_heap`.
+    /// @notice Returns the number of users in the `_heap`.
     /// @param _heap The heap parameter.
-    /// @return The size of the heap.
+    /// @return The length of the heap.
     function getSize(Heap storage _heap) internal view returns (uint256) {
         return _heap.accounts.length;
     }
```

Comment:
Should we use this function instead of `_heap.accounts.length` each time?

Reply:
I'm not sure I understand
you're proposing to use `getSize` instead of `_heap.accounts.length` throughout the entire code?

Reply:
Yes that's the idea. Just to make the code more concise, but not sure it's a good idea tbh

Reply:
Do we agree on not implementing this one?